keps/sig-network/20200103-mixed-protocol-lb.md

```diff
@@ -0,0 +1,372 @@
+  name: mixed-protocol
+spec:
+  type: LoadBalancer
+  ports:
+    - name: dns-udp
+      port: 53
+      protocol: UDP
+    - name: dns-tcp
+      port: 53
+      protocol: TCP
+  selector:
+    app: my-dns-server
+  ```
+
+Once that limit is removed those Service definitions will reach the Cloud Provider LB controller implementations. The logic of the particular Cloud Provider LB controller and of course the actual capabilities and architecture of the backing Cloud Load Balancer services determines how the actual exposure of the application really mnaifests. For this reason it is important to understand the capabilities of those backing services and to design this feature accordingly.
+
+### User Stories [optional]
+
+#### Story 1
+
+As a Kubernetes cluster user I want to expose my application that provides its services on different protocols on a single cloud provider load balancer IP. In order to achieve this I want to define different `ports` with different protocols in my Service definition with `type: LoadBalancer`
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Alibaba
+
+The Alibaba CPI supports TCP, UDP and HTTPS in Service definitions and can configure the SLB listeners with the protocols defined in the Service.
+
+The Alibaba SLB supports TCP, UDP and HTTPS listeners, an own listener must be configured for each protocol, and those listeners can be assigned to the same SLB instance. 
+
+The number of listeners does not affect SLB pricing.
+https://www.alibabacloud.com/help/doc-detail/74809.htm
+
+A user can ask for an internal TCP/UDP Load Balancer via a K8s Service definition that also has the annotation `service.beta.kubernetes.io/alibaba-cloud-loadbalancer-address-type: "intranet"`. Internal SLBs are free.
+
+Summary: Alibaba CPI and SLB seems to support mixed protocol Services, and the pricing is not based on the number of protocols per Service.
+
+#### AWS
+
+The AWS CPI does not support other protocol in the Service definition than TCP for load balancers. Beside that the AWS CPI looks for annotations on the Service to determine whether TCP, TLS or HTTP(S) listener should be created in the AWS ELB for a configured Service port.
+
+AWS Classic LB supports TCP,TLS, and HTTP(S) protocols behind the same IP address. 
+
+AWS Network LB supports TCP/TLS and UDP protocols behind the same IP address. As we can see, UDP cannot be utilized currently, due to the limitation in the AWS CPI.
+
+If NLB is used the HTTP(S) support would require AWS Application LB, but on the other hand that LB type does not support TCP or UDP, i.e. the usage of TCP+HTTP or UDP+HTTP on the same LB instace behind the same IP address is not possible in AWS.
+
+From pricing perspective the AWS NLB and the CLB have the following models:
+https://aws.amazon.com/elasticloadbalancing/pricing/
+Both are rather usage based than "number of protocols" based, though it is true that in case of NLB TCP, UDP and TLS has separated quotas in the pricing units.
```

Comment:
```suggestion
Both are primarily usage based rather than charging based on the number of protocol, however NLBs have separate pricing unit quotas for TCP, UDP and TLS.
```